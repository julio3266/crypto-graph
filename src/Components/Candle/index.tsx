import React, { useRef } from 'react';
import { scaleLinear } from 'd3';
import Theme from '@Global/Theme';
import Svg from 'react-native-svg';
import { Dimensions, ScrollView, View } from 'react-native';
import { Candles, CandlesModel } from './Candles';
import Animated, {
    useSharedValue,
    useAnimatedStyle,
    useAnimatedGestureHandler,
} from 'react-native-reanimated';
import {
    PanGestureHandler,
    GestureHandlerRootView,
    PanGestureHandlerGestureEvent,
} from 'react-native-gesture-handler';
import { useChart } from '@Src/hooks/useChart';

const { heightScreen } = Theme.dimensions;
export const { width: windowWidth } = Dimensions.get('window');


const MIN_CANDLE_WIDTH_CM = 8;
const MIN_CANDLE_WIDTH_PX = MIN_CANDLE_WIDTH_CM / 2;
const MAX_CANDLE_HEIGHT_PX = heightScreen / 5;

export const Candle: React.FC = () => {

    const scrollViewRef = useRef<ScrollView>(null);
    const candlestickData = useChart();
    const scale = useSharedValue(1);
    const translateX = useSharedValue(0);
    const translateY = useSharedValue(0);

    const getDomain = (rows: CandlesModel[]): [number, number] => {
        const values = rows.map(({ high, low }) => [high, low]).flat();
        return [Math.min(...values), Math.max(...values)];
    };

    const domain = getDomain(candlestickData);
    const candleWidth = Math.max(windowWidth / 40, MIN_CANDLE_WIDTH_PX);
    const width = candleWidth * scale.value;
    const scaleY = scaleLinear()
        .domain(domain)
        .range([MAX_CANDLE_HEIGHT_PX, 0]);
    const scaleBody = scaleLinear()
        .domain([1, Math.max(...domain) - Math.min(...domain)])
        .range([0, MAX_CANDLE_HEIGHT_PX]);

    const panHandler = useAnimatedGestureHandler<PanGestureHandlerGestureEvent>({
        onActive: (event) => {
            translateX.value = event.translationX;
            translateY.value = event.translationY;
        }
    });

    const animatedStyle = useAnimatedStyle(() => ({
        transform: [
            { scale: scale.value },
            { translateX: translateX.value },
            { translateY: translateY.value },
        ],
    }));

    return (
        <GestureHandlerRootView style={{ flex: 1 }}>
            <PanGestureHandler onGestureEvent={panHandler}>
                <Animated.View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
                    <ScrollView ref={scrollViewRef} horizontal contentContainerStyle={{ justifyContent: 'center', alignItems: 'center' }}>
                        <Animated.View style={[animatedStyle, { justifyContent: 'center', alignItems: 'center' }]}>
                            <Svg width={Math.max(candlestickData.length * width, windowWidth)} height={heightScreen / 2}>
                                {candlestickData.map((candle, index) => (
                                    <Candles
                                        key={index}
                                        candle={candle}
                                        index={index}
                                        width={width}
                                        scaleY={scaleY}
                                        scaleBody={scaleBody}
                                    />
                                ))}
                            </Svg>
                        </Animated.View>
                    </ScrollView>
                </Animated.View>
            </PanGestureHandler>
        </GestureHandlerRootView>
    );
};
