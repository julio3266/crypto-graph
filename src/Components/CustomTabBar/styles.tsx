import Styled from 'styled-components/native'

export const Container = Styled.View`
    background-color: #000;
    height: 10%;
`