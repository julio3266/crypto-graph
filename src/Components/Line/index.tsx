import React, { useEffect, useRef, useState } from 'react';
import {
    Canvas,
    Path,
    Circle,
    SkPath,
    Skia,
    useValue,
    runTiming,
    useComputedValue,
    Text,
    useFont,
} from '@shopify/react-native-skia';
import Theme from '@Global/Theme';
import { area, line, scaleLinear, scaleTime } from 'd3';
import { Easing } from 'react-native';

export const SOCKET_URL = "wss://nbstream.binance.com/eoptions/ws";
const GRAPH_HEIGHT = 250;
const GRAPH_WIDTH = Theme.dimensions.widthScreen;

export interface DataPoint {
    date: string;
    value: number;
}

export interface GraphData {
    min: number;
    max: number;
    curve: SkPath;
    area: SkPath;
}

export const Line: React.FC = ({ }) => {
    const transition = useValue(1);
    const state = useValue({
        current: 0,
        next: 1,
    });

    const [data, setData] = useState<DataPoint[]>([]);
    const [latestDataPoint, setLatestDataPoint] = useState<DataPoint | null>(null);
    const font = useFont(require("@Src/Assets/Roboto-Light.ttf"), 12);

    const makeGraph = (data: DataPoint[]): GraphData => {
        const max = Math.max(...data.map(val => val.value));
        const min = Math.min(...data.map(val => val.value));
        const y = scaleLinear().domain([min, max]).range([GRAPH_HEIGHT, 35]);

        const x = scaleTime()
            .domain([new Date(data[0].date), new Date(data[data.length - 1].date)])
            .range([10, GRAPH_WIDTH - 10]);

        const straightLine = line<DataPoint>()
            .x(d => x(new Date(d.date)))
            .y(d => y(d.value))(data);

        const filledArea = area<DataPoint>()
            .x(d => x(new Date(d.date)))
            .y0(GRAPH_HEIGHT)
            .y1(d => y(d.value))(data);

        const skLinePath = Skia.Path.MakeFromSVGString(straightLine!);
        const skAreaPath = Skia.Path.MakeFromSVGString(filledArea!);

        return {
            max,
            min,
            curve: skLinePath!,
            area: skAreaPath!,
        };
    };

    const transitionStart = () => {
        state.current = {
            current: state.current.next,
            next: (state.current.next + 1) % 2,
        };
        transition.current = 0;
        runTiming(transition, 1, {
            duration: 750,
            easing: Easing.inOut(Easing.cubic),
        });
    };

    const graphData = data.length > 0 ? [makeGraph(data)] : [];

    const path = useComputedValue(() => {
        if (graphData.length > 0) {
            const start = graphData[0].curve;
            const end = graphData[0].curve;
            const result = start.interpolate(end, transition.current);
            return result?.toSVGString() ?? '0';
        }
        return '0';
    }, [state, transition, graphData]);

    const areaPath = useComputedValue(() => {
        if (graphData.length > 0) {
            const start = graphData[0].area;
            const end = graphData[0].area;
            const result = start.interpolate(end, transition.current);
            return result?.toSVGString() ?? '0';
        }
        return '0';
    }, [state, transition, graphData]);

    const handleNewData = (newData: DataPoint) => {
        setData((prevData) => {
            const updatedData = [...prevData, newData].slice(-30);
            transitionStart();
            return updatedData;
        });
        setLatestDataPoint(newData);
    };

    const dataBuffer = useRef<DataPoint[]>([]);

    useEffect(() => {
        const ws = new WebSocket(SOCKET_URL);

        ws.onopen = () => {
            ws.send(JSON.stringify({
                method: "SUBSCRIBE",
                params: ["ETHUSDT@index"],
                id: 1
            }));
        };

        ws.onmessage = (event) => {
            const data = JSON.parse(event.data);
            if (data.e === 'index') {
                dataBuffer.current.push({ date: new Date().toISOString(), value: parseFloat(data.p) });
            }
        };

        ws.onclose = () => {

        };

        ws.onerror = (error) => {

        };

        const intervalId = setInterval(() => {
            if (dataBuffer.current.length > 0) {
                handleNewData(dataBuffer.current[dataBuffer.current.length - 1]);
                dataBuffer.current = [];
            }
        }, 3000);

        return () => {
            ws.close();
            clearInterval(intervalId);
        };
    }, []);
    if (!font) {
        return null
    }
    return (
        <Canvas
            style={{
                width: GRAPH_WIDTH,
                height: GRAPH_HEIGHT,
            }}>
            {graphData.length > 0 && (
                <>
                    <Path path={areaPath} color={`${Theme.colors.pink}55`} />
                    <Path style="stroke" path={path} strokeWidth={2} color={Theme.colors.pink} />
                    {latestDataPoint && (
                        <>
                            <Circle
                                cx={scaleTime().domain([new Date(data[0].date), new Date(data[data.length - 1].date)]).range([10, GRAPH_WIDTH - 10])(new Date(latestDataPoint.date))}
                                cy={scaleLinear().domain([Math.min(...data.map(val => val.value)), Math.max(...data.map(val => val.value))]).range([GRAPH_HEIGHT, 35])(latestDataPoint.value)}
                                r={4}
                                color={Theme.colors.pink}
                            />
                            <Text
                                x={scaleTime().domain([new Date(data[0].date), new Date(data[data.length - 1].date)]).range([10, GRAPH_WIDTH - 30])(new Date(latestDataPoint.date))}
                                y={scaleLinear().domain([Math.min(...data.map(val => val.value)), Math.max(...data.map(val => val.value))]).range([GRAPH_HEIGHT, 20])(latestDataPoint.value) - 10}
                                text={latestDataPoint.value.toFixed(2)}
                                font={font}
                                color={Theme.colors.white}
                            />

                        </>
                    )}
                </>
            )}
        </Canvas>

    );
};
