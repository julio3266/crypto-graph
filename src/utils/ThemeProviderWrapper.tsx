import React from "react";
import { ThemeProvider } from "styled-components/native";
import theme from "../Global/Theme";

interface IProps {
    children: any;
    padding?: boolean;
}

export const ThemeProviderWrapper: React.FC<IProps> = ({ children }) => {
    return (<ThemeProvider theme={theme}>{children}</ThemeProvider>
    );
};