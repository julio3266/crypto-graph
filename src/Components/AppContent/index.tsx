import React from 'react';
import { Container } from './styles'

export interface AppContentProps {
    children: React.ReactNode
}

export const AppContent: React.FC<AppContentProps> = ({ children }) => {
    return (
        <Container>
            {children}
        </Container>
    )
}