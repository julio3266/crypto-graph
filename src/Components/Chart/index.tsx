import React from 'react'
import { ChartContainer } from './styles'
import { Line } from '../Line'

export type chartType = 'Line' | 'Candle'

interface ChartProps {
  type: chartType
}

export const Chart: React.FC<ChartProps> = ({ type }) => {
  return (
    <ChartContainer>
      <Line />
    </ChartContainer>
  )
}
