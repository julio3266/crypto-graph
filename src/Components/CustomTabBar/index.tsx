import React from "react";
import { AnimatedView, Btn, Container, HomeIcon, SearchIcon, Title, WalletIcon } from './styles'

import { assignTestId } from "@Src/utils/QualityAssurance";
import { NavigationHelpers, ParamListBase, TabNavigationState, useNavigation } from "@react-navigation/native";
import { BottomTabDescriptorMap, BottomTabNavigationEventMap } from "@react-navigation/bottom-tabs/lib/typescript/src/types";
import { Animated } from "react-native";

export interface CustomTabBarProps {
    testID?: string;
    state: TabNavigationState<ParamListBase>;
    descriptors: BottomTabDescriptorMap;
    navigation: NavigationHelpers<ParamListBase, BottomTabNavigationEventMap>;
}

export const CustomTabBar: React.FC<CustomTabBarProps> = ({
    testID = 'CustomTabBarID',
    state,
    descriptors,
    navigation
}) => {
    return (
        <Container {...assignTestId("View", testID)}>
            {
                state?.routes.map((route, index: number) => {
                    const opacity = React.useRef(new Animated.Value(1)).current;
                    const isFocused = state.index === index;

                    const onPress = () => {

                        const event = navigation.emit({
                            type: 'tabPress',
                            target: route.key,
                            canPreventDefault: true
                        });

                        if (!isFocused && !event.defaultPrevented) {
                            navigation.navigate(route.name);
                        }
                    };

                    return (
                        <AnimatedView
                            {...assignTestId('Lottie', `${testID}_tabBarAnimatedView_${index}`)}
                            key={index}
                            style={[
                                {
                                    opacity: isFocused ? 1 : 0.5,
                                    transform: [
                                        { scaleX: isFocused ? 1 : 1 },
                                        { scaleY: isFocused ? 1 : 1 },
                                    ],
                                },
                            ]}
                        >
                            <Btn
                                isFocused={isFocused}
                                onPress={() => onPress()}
                                {
                                ...assignTestId('TouchableOpacity', `${testID}_tabBarBtn_${index}`)
                                }
                            >
                                {
                                    route.name === 'Home' && (
                                        <>
                                            <HomeIcon isFocused={isFocused} />
                                            {isFocused && (<Title>{route.name}</Title>)}
                                        </>
                                    ) ||
                                    route.name === 'Search' && (
                                        <>
                                            <SearchIcon isFocused={isFocused} />
                                            {isFocused && (<Title>{route.name}</Title>)}
                                        </>
                                    ) ||
                                    route.name === 'Wallet' && (
                                        <>
                                            <WalletIcon isFocused={isFocused} />
                                            {isFocused && (<Title>{route.name}</Title>)}
                                        </>
                                    )
                                }
                            </Btn>
                        </AnimatedView>
                    )
                })
            }
        </Container>
    );

}