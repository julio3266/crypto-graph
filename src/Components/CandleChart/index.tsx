import React, { useEffect, useState } from "react";
import { Container, Content } from "./styles";
import { useChart } from '@Src/hooks/useChart';
import { StyleSheet, View } from "react-native";
import { PanGestureHandler, State } from "react-native-gesture-handler";
import { onGestureEvent, useValues } from 'react-native-redash';
import Animated, {
    add,
    diffClamp,
    eq,
    modulo,
    sub,
} from "react-native-reanimated";

import { Header } from "@Src/Components/CandleChart/Header";
import { Candle } from "@Src/Components/CandleChart/Candle";
import { size, Chart } from "@Src/Components/CandleChart/Chart";
import { Line } from "@Src/Components/CandleChart/Line";


const getDomain = (rows: Candle[]): [number, number] => {
    const values = rows.map(({ high, low }) => [high, low]).flat();
    return [Math.min(...values), Math.max(...values)];
};

export const CandleChart = () => {
    const candlestickData = useChart();
    const [x, y, state] = useValues(0, 0, State.UNDETERMINED);
    const [startIndex, setStartIndex] = useState(0);
    const [candles, setCandles] = useState<Candle[]>([]);

    useEffect(() => {
        const newCandles = candlestickData.slice(startIndex, startIndex + 40);
        setCandles(newCandles);
    }, [startIndex, candlestickData]);

    const loadMoreCandles = () => {
        setStartIndex(prevIndex => prevIndex + 2);
    };

    useEffect(() => {
        if (candles.length === 20) {
            loadMoreCandles();
        }
    }, [candles]);

    const domain = getDomain(candles);

    const gestureHandler = onGestureEvent({
        x,
        y,
        state,
    });

    const caliber = size / candles.length;
    const translateY = diffClamp(y, 0, size);
    const translateX = add(sub(x, modulo(x, caliber)), caliber / 2);
    const opacity = eq(state, State.ACTIVE);

    return (
        <Container >
            <Content>
                <Header candles={candles} />
                <View>
                    {candles.length > 1 && (
                        <>
                            <Chart {...{ candles, domain }} />
                            <PanGestureHandler minDist={0} {...gestureHandler}>
                                <Animated.View style={StyleSheet.absoluteFill}>
                                    <Animated.View
                                        style={{
                                            transform: [{ translateY }],
                                            opacity,
                                            ...StyleSheet.absoluteFillObject,
                                        }}
                                    >
                                        <Line x={size} y={0} />
                                    </Animated.View>
                                    <Animated.View
                                        style={{
                                            transform: [{ translateX }],
                                            opacity,
                                            ...StyleSheet.absoluteFillObject,
                                        }}
                                    >
                                        <Line x={0} y={size} />
                                    </Animated.View>
                                </Animated.View>
                            </PanGestureHandler>
                        </>
                    )}
                </View>
            </Content>
        </Container>
    );
};
