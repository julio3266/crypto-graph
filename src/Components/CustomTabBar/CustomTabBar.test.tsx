import React from "react";
import { render } from "@testing-library/react-native";
import { mockTestID } from "@Src/utils/QualityAssurance";
import { ThemeProviderWrapper as wrapper } from "@Src/utils/ThemeProviderWrapper";
import { CustomTabBar, CustomTabBarProps } from "@Src/Components/CustomTabBar";

describe("Components/CustomTabBar", () => {
    let props: CustomTabBarProps = {
        testID: "CustomTabBarID",
        descriptors: 'teste',
        navigation: 'teste',
        state: 'teste'
    };

    it(`should render CustomTabBar component`, async () => {
        const { findByTestId, toJSON } = render(<CustomTabBar {...props} />, {
            wrapper,
        });

        const sut_container = mockTestID("View", props.testID!).testID;
        const el_btn = mockTestID("TouchableOpacity", `${props.testID!}_tabBarBtn`).testID;
        const el_btn_title = mockTestID("Text", `${props.testID!}_tabBarBtnTitle`).testID;

        const sut = await findByTestId(sut_container);
        expect(sut).toBeTruthy();

        const button = await findByTestId(el_btn);
        expect(button).toBeTruthy();

        const buttonTitle = await findByTestId(el_btn_title);
        expect(buttonTitle).toBeTruthy();

        expect(toJSON()).toMatchSnapshot();
    });
});
