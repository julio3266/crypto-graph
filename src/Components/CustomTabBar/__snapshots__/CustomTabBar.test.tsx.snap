// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/CustomTabBar should render CustomTabBar component 1`] = `
<View
  accessible={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#262250",
      "borderTopLeftRadius": 30,
      "flexDirection": "row",
      "height": "9%",
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
  testID="View_CustomTabBarID"
>
  <View
    accessible={false}
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": "35%",
        "justifyContent": "center",
        "opacity": 1,
        "transform": [
          {
            "scaleX": 1,
          },
          {
            "scaleY": 1,
          },
        ],
      }
    }
    testID="Lottie_CustomTabBarID_tabBarAnimatedView_0"
  >
    <View
      accessible={false}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FF409A",
          "borderBottomLeftRadius": 60,
          "borderBottomRightRadius": 60,
          "borderTopLeftRadius": 60,
          "borderTopRightRadius": 60,
          "height": 55,
          "justifyContent": "center",
          "opacity": 1,
          "width": 98,
        }
      }
      testID="TouchableOpacity_CustomTabBarID_tabBarBtn_0"
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={30}
        bbWidth={30}
        color={4294967295}
        fill="none"
        focusable={false}
        height={30}
        isFocused={true}
        meetOrSlice={0}
        minX={0}
        minY={0}
        stroke="#fff"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {},
            {
              "flex": 0,
              "height": 30,
              "width": 30,
            },
          ]
        }
        tintColor={4294967295}
        vbHeight={24}
        vbWidth={24}
        width={30}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={4294967295}
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        >
          <RNSVGPath
            d="m3 9 9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
          />
          <RNSVGPath
            d="M9 22V12h6v10"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 14,
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Home
      </Text>
    </View>
  </View>
  <View
    accessible={false}
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": "35%",
        "justifyContent": "center",
        "opacity": 0.5,
        "transform": [
          {
            "scaleX": 1,
          },
          {
            "scaleY": 1,
          },
        ],
      }
    }
    testID="Lottie_CustomTabBarID_tabBarAnimatedView_1"
  >
    <View
      accessible={false}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
          "borderTopLeftRadius": 0,
          "borderTopRightRadius": 0,
          "height": 45,
          "justifyContent": "center",
          "opacity": 1,
          "width": 98,
        }
      }
      testID="TouchableOpacity_CustomTabBarID_tabBarBtn_1"
    />
  </View>
</View>
`;
